app/services/payout_report_publisher_includer.rb

```diff
@@ -0,0 +1,57 @@
+class PayoutReportPublisherIncluder < BaseService
+  def initialize(payout_report:, publisher:, should_send_notifications:)
+    @publisher = publisher
+    @payout_report = payout_report
+    @should_send_notifications = should_send_notifications
+  end
+
+  def perform
+    return if @publisher.suspended? || !@publisher.has_verified_channel?
+    publisher_has_unsettled_balance = false
+
+    wallet = @publisher.wallet
+    return if wallet.nil?
+
+    probi = wallet.channel_balances[@publisher.owner_identifier].probi_before_fees # probi = balance
+    if probi.positive?
+      publisher_has_unsettled_balance = true
+
+      if @publisher.uphold_verified? && wallet.address.present?
+        PotentialPayment.create(payout_report_id: @payout_report.id,
```

Comment:
Try to lock everything in a transaction: I'm afraid if this function fails after this block, we'd get an extra `PotentialPayment`. Preferably we refactor this class so lookups are the in the beginning, and writes are at the end.

Reply:
Hm I don't think we want a transaction here since that would result in a publisher not being paid at all rather them being paid of only some of their channels or referrals, but maybe I'm misunderstanding your comment.

Which line are you worried about failing, and what result would that produce?

Reply:
So let's say this service fails in the middle (perhaps during a database lookup). `IncludePublisherInPayoutReportJob` will restart and call `PayoutReportPublisherIncluder`.

In finer detail: If `PotentialPayment` for Referrals on line 20 succeeds, and let's say there's a failure while `@publisher.channels.verified.each` is running for PotentialPayment's Contributions, then the retried `IncludePublisherInPayoutReportJob` will have duplicates for (1) `PotentialPayment` for Referrals (2) `PotentialPayment` for Contributions from the previously failed job.

Reply:
Thanks for the clarification.  That's a valid concern and one that I try to address in the PotentialPayment validations.

Part of the design goal of the refactor was to make it so you could run the job multiple times and not add any duplicate payments.  That way if a publisher was left out for whatever reason, we could just run the job again and supply the publisher (or just re run for all publishers) and include them without fear of a duplicate payment.

The validations are set up so that a channel can have at most one payment for a given payout report, https://github.com/brave-intl/publishers/pull/1380/files#diff-31551a82b5c4e5a8bcf61f64060eb102R12, and a publisher can only receive one referral payment for a given payout report https://github.com/brave-intl/publishers/pull/1380/files#diff-31551a82b5c4e5a8bcf61f64060eb102R18.

The tests for validations might also help demonstrate what I'm after https://github.com/brave-intl/publishers/pull/1380/files#diff-1db04edc63db4ff7684664f9115af41aR3.